content/courses/Ada_For_The_C_Embedded_Developer/chapters/04_Embedded.rst

```diff
@@ -240,9 +240,194 @@ Dealing with Absence of FPU with Fixed Point
+    {
+        volatile double val = 0.0;
+        int i;
+
+        for (i = 0; i < 1000; i++)
+        {
+            val += i * 2.0;
+        }
+        printf ("val: %5.3f\n", val);
+    }
+
+In this example, :c:`val` has the modifier :c:`volatile`, which indicates that
+the compiler must handle :c:`val` as a volatile object. Therefore, each read
+and write access in the loop is performed by accessing the value of :c:`val` in
+then memory.
+
+This is the corresponding implementation in Ada:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Embedded.Volatile_Object_Ada
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Volatile_Object is
+       Val : Long_Float with Volatile;
+    begin
+       Val := 0.0;
+       for I in 0 .. 999 loop
+          Val := Val + 2.0 * Long_Float (I);
+       end loop;
+
+       Put_Line ("Val: " & Long_Float'Image (Val));
+    end Show_Volatile_Object;
+
+In this example, :ada:`Val` has the :ada:`Volatile` aspect, which makes the
+object volatile. We can also use the :ada:`Volatile` aspect in type
+declarations. For example:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Embedded.Volatile_Type
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Volatile_Type is
+       type Volatile_Long_Float is new Long_Float with Volatile;
+
+       Val : Volatile_Long_Float;
+    begin
+       Val := 0.0;
+       for I in 0 .. 999 loop
+          Val := Val + 2.0 * Volatile_Long_Float (I);
+       end loop;
+
+       Put_Line ("Val: " & Volatile_Long_Float'Image (Val));
+    end Show_Volatile_Type;
+
+Here, we're declaring a new type :ada:`Volatile_Long_Float` based on the
+:ada:`Long_Float` type and using the :ada:`Volatile` aspect. Any object of this
+type is automatically volatile.
+
+In addition to that, we can declare components of an array to be volatile. In
+this case, we can use the :ada:`Volatile_Components` aspect in the array
+declaration. For example:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Embedded.Volatile_Array_Components
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Volatile_Array_Components is
+       Arr : array (1 .. 2) of Long_Float with Volatile_Components;
+    begin
+       Arr := (others => 0.0);
+
+       for I in 0 .. 999 loop
+          Arr (1) := Arr (1) +  2.0 * Long_Float (I);
+          Arr (2) := Arr (2) + 10.0 * Long_Float (I);
+       end loop;
+
+       Put_Line ("Arr (1): " & Long_Float'Image (Arr (1)));
+       Put_Line ("Arr (2): " & Long_Float'Image (Arr (2)));
+    end Show_Volatile_Array_Components;
+
+Note that it's possible to use the :ada:`Volatile` aspect for the array
+declaration as well:
+
+[Ada]
+
+.. code-block:: ada
+
+    Arr : array (1 .. 2) of Long_Float with Volatile;
+
+Atomic
+~~~~~~
+
+Operations are said to be atomic when they can be completed without
+interruptions. This is an important requirement when we're performing
```

Comment:
Regarding operations, that's a different use of the term "atomic" (eg, for a database update operation) than what the RM is talking about.  Here, atomic means that any read or update of such an object must be done in the generated assembly language such that no other instruction could execute on that same object before the read or update completes. (That's not the right wording, but that's the idea.) So generally that means read or update with one assembly language instruction. For example, if we have a 32-bit object composed of four 8-bit bytes, the compiler cannot generate code to read or update the object using four 8-bit store/load instructions, or even two 16-bit store/load instructions. 
That means that there are objects for which the Atomic aspect cannot be specified. Lots of machines support integer types that are larger than the native word-sized integer. For example, a 16-bit machine probably supports both 16-bit and 32-bit integers, but only 16-bit integer objects can be marked as atomic. And more generally, only objects that fit into at most 16-bits, on such a machine.

Reply:
Thanks for this explanation; it is very good! Actually, my understanding was the same as yours, but I wasn't using the correct terminology.

I decided to use your explanation in an admonition, so the context becomes clearer to the readers.